Add --enable-small to FFmpeg for size optimized builds

Nomis101 opened the pull request:
See #2571 for details.


**Description of Change:**
`--optimize=size` should trigger FFmpeg to build with `--enable-small`. This reduces the size of the application about 9%, as my tests have shown on macOS 10.15.





**Test on:**
- [x] macOS 10.15
- [x] Linux
- [x] Windows (mingw-w64)

bradleysepos commented:
For clang/macOS, I'd prefer for `--enable-small` to only be enabled given `--optimize=size-aggressive` or a similar new value to invoke `-Oz` globally, to avoid mixing `-Os` and `-Oz`. My reasoning per #2571 being that `-Os` is generally considered not to harm performance, whereas `-Oz` is allowed to harm performance in order to reduce binary size.

Edit: Perhaps I can be a bit more clear. When compiling with `--optimize=size` and thus `-Os`, one can assume performance will not be significantly impacted, if at all. This would potentially change that behavior.

Nomis101 commented:
OK, I see. I will rework the PR accordingly.

Nomis101 commented:
`--optimize=size-aggressive` now triggers `--enable-small` for FFmpeg and switches from `-Os` to `-Oz`. But actually, HandBrake/macosx is still build with `-Os`, but I couldn't find why.

bradleysepos commented:
Looking good, thanks. I suspect we should add a compiler test for `-Oz`, only print it in the configure help if it's available, and error out if specified but unavailable. A bit overkill but I can probably work on it if it's too much trouble.

It looks like `-Os` is set in the Xcode project. We can make this configurable with a variable but perhaps that should be an additional PR. @galad87 Any thoughts on this?

Nomis101 commented:
> Looking good, thanks. I suspect we should add a compiler test for `-Oz`, only print it in the configure help if it's available, and error out if specified but unavailable. A bit overkill but I can probably work on it if it's too much trouble.

Yes, this really seems a bit tricky to implement. Currently I'm a bit lost how to do so.
But I guess it has to be done similar then this one: https://github.com/HandBrake/HandBrake/commit/53d6226c58e9f7bba23f1746bad1d8b61ec6338d

Nomis101 commented:
Not sure, if you had this in mind, but it seems to work. Now I need to think about the part "error out if specified but unavailable".

bradleysepos commented:
That's the way to approach it. Configure will error out if an invalid value is passed for optimize. Need to move the check after the tools are re-run for cross compilation, though. Just move the entire debugMode and optimizeMode sections after the `if cross:` block.

bradleysepos commented:
Looks good, tested. Will merge soon. Thanks!

Nomis101 committed:
contrib: Allow optimizing x265 with -Oz via --optimize=size-aggressive.

Additionally reduces file size 4.5%.

bradleysepos commented:
It looks like `MinSizeRel` invokes `-Os`, not `-Oz`?

bradleysepos commented:
To be clear, I think it should probably be enabled for both `--optimize=size` and `--optimize=aggressive`, assuming it's not overriding `-Oz`.

Nomis101 commented:
> It looks like `MinSizeRel` invokes `-Os`, not `-Oz`?

Yes, `MinSizeRel` sets the gcc compiler optimization level to `-Os` whereas `Release`
sets it to `-O3`. I don't think it can be made to use `-Oz`, seems to be cmake related.
So, you think `MinSizeRel` should also be enabled for `optimize=size`? Its not overriding `-Oz`, I checked that.

bradleysepos commented:
Yes, since `size` == `-Os`.

Nomis101 commented:
> Yes, since `size` == `-Os`.

OK, I will change that, after I fixed the issue of my Xcode not starting anymore on Catalina, since updated to the newest version.

Nomis101 committed:
build: Add --optimize=size-aggressive (-Oz).

Compiles with -Oz. -Oz is like -Os but further reduces total size even at the expense of speed.

Nomis101 committed:
contrib: Configure FFmpeg with --enable-small when using --optimize=size-aggressive.

Closes #2571.

bradleysepos commented:
Consolidated into 3 commits and merged. Thanks!

Nomis101 commented:
Thanks! I ever mess up the PR if trying to consolidate, so thanks for doing this.
